include/mbedtls/ssl.h

```diff
@@ -2266,6 +2304,132 @@ void mbedtls_ssl_set_export_keys_cb( mbedtls_ssl_context *ssl,
                                      mbedtls_ssl_export_keys_t *f_export_keys,
                                      void *p_export_keys );
 
+/** \brief Set the user data in an SSL configuration to a pointer.
+ *
+ * You can retrieve this value later with mbedtls_ssl_conf_get_user_data_p().
+ *
+ * \note The library stores \c p without accessing it. It is the responsibility
+ *       of the caller to ensure that the pointer remains valid.
+ *
+ * \param conf           The SSL configuration context to modify.
+ * \param p              The new value of the user data.
+ */
+static inline void mbedtls_ssl_conf_set_user_data_p(
+    mbedtls_ssl_config *conf,
+    void *p )
+{
+    conf->MBEDTLS_PRIVATE(user_data) = (uintptr_t) p;
+}
+
+/** \brief Set the user data in an SSL configuration to an integer.
+ *
+ * You can retrieve this value later with mbedtls_ssl_conf_get_user_data_n().
+ *
+ * \param conf           The SSL configuration context to modify.
+ * \param n              The new value of the user data.
+ */
+static inline void mbedtls_ssl_conf_set_user_data_n(
+    mbedtls_ssl_config *conf,
+    uintptr_t n )
+{
+    conf->MBEDTLS_PRIVATE(user_data) = n;
+}
+
+/** \brief Retrieve the user data in an SSL configuration as a pointer.
+ *
+ * This is the value last set with mbedtls_ssl_conf_set_user_data_n(), or
```

Comment:
The comments appear to be reversed for mbedtls_ssl_conf_get_user_data_p() and mbedtls_ssl_conf_get_user_data_n().

For an embedded library, is it necessary to have both _user_data_p() and _user_data_n() accessors instead of a single _user_data() for a (void *) pointer?  The caller can cast to (uintptr_t) to store an integral value instead of a pointer.

Reply:
I expect most users to use pointers, which is why I provide pointer accessors. But I've been burnt in the past with similar interfaces that had a pointer user data and I wanted to store a handle: there's no portable way to store an integer in a pointer variable. Hence I'm providing integer accessors.

Accessors are static inline functions so they don't cost anything if you don't use them.

Reply:
> here's no portable way to store an integer in a pointer variable.

https://en.cppreference.com/w/c/types/integer
C99 defines uintptr_t and intptr_t to be at least sizeof(void *).

*`mbedtls_ssl_conf_set_user_data_n` and `mbedtls_ssl_set_user_data_n` use `uintptr_t`, so I do not understand how portability of casting through (uintptr_t) is even a question*. (e.g. `(void *)(uintptr_t)n`)

As I pointed out in #5413, similar interfaces in openssl and gnutls already exist: openssl `SSL_get_app_data()` and gnutls `gnutls_session_get_ptr()`.  They operate on `(void *)`, and do not have a separate `(uintptr_t)` interface.

I like that mbedtls is usually lean with its interfaces, unlike openssl which has waaaay too many public functions.  The extra interfaces here for _p and _n feel excessive to me when a single interface taking `(void *)` would suffice, IMHO.

Reply:
It is portable to cast a `void*` to `uintptr_t` and back, assuming that the C implementation has `uintptr_t` at all. The C standard doesn't guarantee anything about the correspondence between integer values and pointer values, but it guarantees that if you put a valid pointer in then you get a pointer to the same object out, and if you put a null poiner in then you get a null pointer out. Mbed TLS makes the assumption that all-bits-zero is both a valid representation of 0 in integer types (because we assume that integer types don't have padding bits) and a valid representation of the null pointer (we don't assume that `NULL` is all-bits-zero, but we do assume `void *p; memset(&p, 0, sizeof(p))); assert(p == NULL);`), which is how the default value of the user data is `0` as an integer and `NULL` as a pointer.

It is not portable to cast a `uintptr_t` to a pointer. This may result in a trap representation of the pointer type, and even just storing a trap representation into a variable has undefined behavior. Architectures where pointers have trap representations are very uncommon today, but they're making a comeback with the push towards security. Storing an arbitrary integer into a pointer allows a security check to say “No, that's not a valid pointer. Abort!”.

Reply:
> It is not portable to cast a uintptr_t to a pointer.

If I understand you correctly (along with other context), then I think that you're saying that it is not portable to cast a `uintptr_t` to a pointer unless that pointer value was what was stored in the `uintptr_t`.  That is, after all, what you have implemented with the `uintptr_t MBEDTLS_PRIVATE(user_data);` member.  Your implementation also assumes that `uintptr_t` will be implemented in environments that use mbedtls, and mbedtls already assumes existence of `uintptr_t`.

There can still be a single getter/setting interface to `user_data`, and you can document undefined behavior if the user calls the getter and casts the result to (void *) if the user had not previously cast that same pointer to `(uintptr_t)` when calling the setter.  My debate above was not whether or not this is undefined behavior, but that there can be a single interface (with fewer interfaces and fewer lines in the header) to portably access the `user_data`.

Reply:
Right, there can't be only `void*` accessors unless we sacrifice functionality or portability. There can be only `uintptr_t` accessors but we'd sacrifice usability.